### PR TITLE
refactor the app source into a reusable field that can be dropped in forms

### PR DESCRIPTION
**Pull request by czue**
update the report builder UI and UCR data source from apps to use this

(note that this is a PR into https://github.com/dimagi/commcare-hq/pull/5608)

@NoahCarnahan I wanted to reuse this widget and it was pretty tricky so ended up going down a fairly deep hole to pull it apart. I think the net result code-wise is an improvement, but curious to get your input.

On the UI side it's currently worse than it was previously (see below). I think it should be pretty simple to restore the old labels, but I wanted to make sure I submitted this before I stopped working on it.

image

### CODE REVIEW

File: corehq/apps/app_manager/fields.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
+import collections
+from django import forms
+from django.utils.translation import ugettext as _
+from corehq.apps.app_manager.models import get_apps_in_domain
+
+
+ApplicationDataSource = collections.namedtuple('ApplicationDataSource', ['application', 'source_type', 'source'])
+
+
+class ApplicationDataSourceWidget(forms.MultiWidget):
```
Comment (czue) [inline corehq/apps/app_manager/fields.py]: bit of django doc digging and debugging to get something workable in these two classes

File: corehq/apps/userreports/reports/builder/forms.py [hunk 2]
```diff
@@ -235,32 +235,13 @@ class CreateNewReportForm(forms.Form):
...
         ],
     )
-
-    application = forms.ChoiceField()
-    source_type = forms.ChoiceField(choices=[
-        ("case", _("Case")),
-        ("form", _("Form")),
-    ])
-    report_source = forms.ChoiceField()
+    report_source = ApplicationDataSourceField()
```
Comment (czue) [inline corehq/apps/userreports/reports/builder/forms.py]: encapsulated by the new field

File: corehq/apps/userreports/reports/builder/forms.py [hunk 3]
```diff
@@ -268,13 +249,7 @@ def __init__(self, domain, *args, **kwargs):
...
                 _('Create New Report'),
                 'report_type',
-                crispy.Field('application', data_bind='value: application'),
-                crispy.Field('source_type', data_bind='value: sourceType'),
-                crispy.Field('report_source', data_bind='''
-                    options: sourcesMap[application()][sourceType()],
-                    optionsText: function(item){return item.text},
-                    optionsValue: function(item){return item.value}
-                '''),
+                crispy.MultiWidgetField('report_source'),
```
Comment (czue) [inline corehq/apps/userreports/reports/builder/forms.py]: same

File: corehq/apps/userreports/templates/userreports/userreports_base.html [hunk 4]
```diff
@@ -35,16 +35,9 @@
         <li>
             <a href="{% url 'create_configurable_data_source_from_app' domain %}">
                 <i class="icon-plus icon-blue"></i>
-                <span class="muted">{% trans "Case data source from application" %}</span>
+                <span class="muted">{% trans "Data source from application" %}</span>
```
Comment (czue) [inline corehq/apps/userreports/templates/userreports/userreports_base.html]: added the case/form selection option, so we only need one view for this now

File: corehq/apps/userreports/ui/forms.py [hunk 5]
```diff
@@ -135,109 +135,21 @@ def save(self, commit=False):
 
 class ConfigurableDataSourceFromAppForm(forms.Form):
 
-    app_id = forms.ChoiceField()
-    case_type = forms.ChoiceField()
+    app_source = ApplicationDataSourceField()
```
Comment (czue) [inline corehq/apps/userreports/ui/forms.py]: this file nicely delgated all the functionality to the field

File: corehq/apps/userreports/views.py [hunk 6]
```diff
@@ -275,33 +263,25 @@ def create_data_source_from_app(request, domain):
...
-    else:
-        form = ConfigurableDataSourceFromAppForm(domain)
-    context = _shared_context(domain)
-    context['data_source_options_map'] = form.data_source_options_map
-    context['form'] = form
-    return render(request, 'userreports/data_source_from_app.html', context)
-
+            app_source = form.cleaned_data['app_source']
+            app = Application.get(app_source.application)
+            if app_source.source_type == 'case':
```
Comment (czue) [inline corehq/apps/userreports/views.py]: combine both views prior logic here

**Comment by czue** [comment]
sorry i didn't break this up more - tried to leave some comments in the PR to explain what's going on

File: corehq/apps/userreports/templates/userreports/data_source_from_app.html [hunk 7]
```diff
@@ -5,12 +5,12 @@
 
 {% block js-inline %}{{ block.super }}
     <script type="text/javascript">
-        $(function(){
-            ko.applyBindings({
-                optionsMap: ko.observable({{ data_source_options_map|JSON }}),
-                appId: ko.observable()
-            }, $('#data-source-config').get(0));
-        });
+        var viewModel = {
```
Comment (benrudolph) [inline corehq/apps/userreports/templates/userreports/data_source_from_app.html]: small, but any reason for taking this outside of the anonymous function? `viewModel` is now a global for the page

**Comment by benrudolph** [comment]
sweet refactor from what i can tell

File: corehq/apps/userreports/templates/userreports/data_source_from_app.html (hunk 7 above)
Comment (czue) [inline corehq/apps/userreports/templates/userreports/data_source_from_app.html]: nope, just copied it from the other file and didn't think about it. https://github.com/dimagi/commcare-hq/commit/beacc5234221a2c75712013313f2793240588666

File: corehq/apps/app_manager/fields.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                sourceType: ko.observable(""),
+                sourcesMap: {{ sources_map|JSON }}
+            }, $("#FORM").get(0));
+        });
+
+    Where FORM is a selector for your form and sources_map is the .all_sources property from this field
+    (which gets set after bootstrap).
+
+    See usages for examples.
+    """
```
Comment (czue) [inline corehq/apps/app_manager/fields.py]: ^ usage

Comment (NoahCarnahan) [inline corehq/apps/app_manager/fields.py]: :heart_eyes: documentation

**Comment by NoahCarnahan** [comment]
Thanks, this looks really good! So satisfying to see net line decreases :) I didn't know about multi widgets before, and this definitely seems like the right use case.

I guess the only remaining inconvenience is needing to remember to add the appropriate knockout bindings. One possibility might be to use the `with` binding to bind our three fields directly to a view model that we construct in the bootstrap method. It would probably make the knockout a bit less readable, and I also think it might prevent a knockout view model bound on the whole form from accessing those fields, which wouldn't be so great. So, that might not actually be an improvement.

Feel free to merge if you think the `with` binding isn't worth pursuing.

**Comment by NoahCarnahan** [comment]
Ha. Didn't see 5796 first. Looking at that now.

**Comment by czue** [comment]
@NoahCarnahan not totally sure i understand what you're saying with the `with` binding, but maybe i'll merge this and let you decide how you want to handle it.

is too bad it wasn't possible/easy to get the UI to look nice with the widget approach - agree would have been nice if that just worked.

**Pull request merged**